pydicom/filereader.py

```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+    IOError
+        When file pointer is not positioned at first byte of first Frame item
+        after Basic Offset Table item or when parsing of Frame item headers
+        fails
+    ValueError
+        When the number of offsets doesn't match the specified number of frames
+
+    """
+    initial_position = fp.tell()
+    offset_values = []
+    current_offset = 0
+    i = 0
+    while True:
+        frame_position = fp.tell()
+        tag = TupleTag(fp.read_tag())
+        if int(tag) == SequenceDelimiterTag:
+            break
+        if int(tag) != ItemTag:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. Expected tag of '
+                f'Frame item #{i} at position {frame_position}.'
+            )
+        length = fp.read_UL()
+        if length % 2:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is not a multiple of 2.'
+            )
+        elif length == 0:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is zero.'
+            )
+
+        first_two_bytes = fp.read(2, 1)
+        if not fp.is_little_endian:
+            first_two_bytes = first_two_bytes[::-1]
+
+        # In case of fragmentation, we only want to get the offsets to the
+        # first fragment of a given frame. We can identify those based on the
+        # JPEG and JPEG 2000 markers that should be found at the beginning and
+        # end of the compressed byte stream.
+        if first_two_bytes in _START_MARKERS:
+            current_offset = frame_position - initial_position
+            offset_values.append(current_offset)
+
+        i += 1
+        fp.seek(length - 2, 1)  # minus the first two bytes
+
+    if len(offset_values) != number_of_frames:
+        raise ValueError(
+            'Number of frame items does not match specified Number of Frames.'
+        )
+    else:
+        basic_offset_table = offset_values
+
+    fp.seek(initial_position, 0)
+    return basic_offset_table
+
+
+class ImageFileReader(object):
+
+    """Reader for DICOM datasets representing Image Information Entities.
+
+    It provides efficient access to individual Frame items contained in the
+    Pixel Data element without loading the entire element into memory.
+
+    Attributes
+    ----------
+    filename: str
+        Path to the DICOM Part10 file on disk
+
+    Examples
+    --------
+    >>> from highdicom.io import ImageFileReader
+    >>> with ImageFileReader('/path/to/file.dcm') as image:
+    ...     print(image.metadata)
+    ...     for i in range(image.number_of_frames):
+    ...         frame = image.read_frame(i)
+    ...         print(frame.shape)
+
+    """
+
+    def __init__(self, filename: str):
+        """
```

Comment:
Can we also include a file object rather than just a filename?  Inevitably someone needing to work with streams will ask for it, although some may be limited if there are negative `seek`s, as seen for example in #753.

Reply:
I generally the idea of allowing the user to alternatively provide an existing file object and have refactored the class accordingly via e1056bd16e193cfd8495cf2e9f690d9fe03cc5a4.

Reply:
I have decided to not set the `is_little_endian` and `is_implicit_VR` attributes on the file object, so that the object does not get modified. Instead an error is raised if their values are not correct.

This is one of the reasons I initially implemented the class such that the file object gets created under the hood and the attributes are set correctly.

@darcymason What are your thoughts on this?